pulser/simulation/simresults.py

```diff
@@ -108,122 +366,136 @@ def get_final_state(self, reduce_to_basis: Optional[str] = None,
+            if self._meas_basis == self._basis_name:
                 # State vector ordered with r first for 'ground_rydberg'
                 # e.g. N=2: [rr, rg, gr, gg] -> [11, 10, 01, 00]
                 # Invert the order ->  [00, 01, 10, 11] correspondence
-                weights = probs if meas_basis == 'digital' else probs[::-1]
+                weights = probs if self._meas_basis == 'digital' \
+                    else probs[::-1]
             else:
-                return {'0' * N: int(N_samples)}
-            weights = weights.flatten()
+                # Only 000...000 is measured
+                weights = np.zeros(probs.size)
+                weights[0] = 1.
 
         elif self._dim == 3:
-            if meas_basis == 'ground-rydberg':
+            if self._meas_basis == 'ground-rydberg':
                 one_state = 0       # 1 = |r>
                 ex_one = slice(1, 3)
-            elif meas_basis == 'digital':
+            elif self._meas_basis == 'digital':
                 one_state = 2       # 1 = |h>
                 ex_one = slice(0, 2)
-
             probs = probs.reshape([3]*N)
-            weights = []
+            weights = np.zeros(2**N)
             for dec_val in range(2**N):
                 ind: list[Union[int, slice]] = []
                 for v in np.binary_repr(dec_val, width=N):
                     if v == '0':
                         ind.append(ex_one)
                     else:
                         ind.append(one_state)
-                # Eg: 'digital' basis => |1> = index 2, |0> = index 0, 1 = 0:2
+                # Eg: 'digital' basis : |1> = index2, |0> = index0, 1 = 0:2
                 # p_11010 = sum(probs[2, 2, 0:2, 2, 0:2])
-                # We sum all probabilites that correspond to measuring 11010,
-                # namely hhghg, hhrhg, hhghr, hhrhr
-                weights.append(np.sum(probs[tuple(ind)]))
+                # We sum all probabilites that correspond to measuring
+                # 11010, namely hhghg, hhrhg, hhghr, hhrhr
+                weights[dec_val] = np.sum(probs[tuple(ind)])
         else:
             raise NotImplementedError(
                 "Cannot sample system with single-atom state vectors of "
-                "dimension > 3."
-                )
-        dist = np.random.multinomial(N_samples, weights)
-        return {np.binary_repr(i, N): dist[i] for i in np.nonzero(dist)[0]}
+                "dimension > 3.")
+        # Takes care of numerical artefacts in case sum(weights) != 1
+        weights /= sum(weights)
+        return cast(ArrayLike, weights)
+
+    def detection_from_basis_state(self, N_d: int, shot: str,
+                                   spam: dict[str, float]) -> Counter:
+        """Computes the distribution of states detected when detecting `shot`.
+
+        Part of the SPAM implementation : computes measurement errors.
+
+        Args:
+            N_d (int): Number of times state has been detected.
```

Comment:
I guess `N_d` should also become `n_d`(or perhaps `n_detections`, it's more informative)